Remora.Rest/Json/DataObjectConverter.cs

```diff
@@ -40,15 +41,15 @@ namespace Remora.Rest.Json;
 /// <typeparam name="TInterface">The interface that is seen in the objects.</typeparam>
 /// <typeparam name="TImplementation">The concrete implementation.</typeparam>
 [PublicAPI]
-public class DataObjectConverter<TInterface, TImplementation> : JsonConverter<TInterface>
+public class DataObjectConverter<TInterface, TImplementation> : JsonConverterFactory
     where TImplementation : TInterface
 {
     private readonly ObjectFactory<TInterface> _dtoFactory;
 
     private readonly IReadOnlyList<PropertyInfo> _dtoProperties;
 
-    // Getters for all properties in the DTO
-    private readonly IReadOnlyDictionary<PropertyInfo, InstancePropertyGetter> _dtoPropertyAccessors;
+    // JSON writers for all properties in the DTO
+    private readonly IReadOnlyDictionary<PropertyInfo, DTOPropertyWriter> _dtoPropertyWriters;
```

Comment:
I didn't think about this earlier; what is the reason for having this as a field on the converter factory, and not the created converter itself?

I would prefer a looser coupling between the two (passing the factory instance to the created converter smells a bit to me).

Reply:
So in case the factory does need to create multiple converters (f.e. due to copying the JsonSerializerOptions and being needed there), we don't need to compile a new set of Expressions for the delegates again.

Either way, the main idea is that the factory does most of the work that would need to be done every time with no changes, and just pass that into the constructor for the actual converters.

Reply:
Got it. Can we pass that data in instead of the converter itself? The created converter just needs the data, not the factory itself even if it happens to be the owner/producer of those data sets.

Reply:
Actually, now that I think about it as well, it is probably better if the logic was moved from the `BoundDataObjectConverter` constructor to the factory's `CreateConverter` and it just passes in the bundled data it needs.

Reply:
Alright, done that.

Take another look over it, mostly I just moved a bit of code around and made members that no longer had to be internal - due to being used just in the `BoundDataObjectConverter` before - private.